src/rest_vol_object.c

```diff
@@ -578,10 +746,368 @@ RV_object_specific(void *obj, const H5VL_loc_params_t *loc_params, H5VL_object_s
+            object_iter_data.op_data                      = args->args.visit.op_data;
+            object_iter_data.is_recursive                 = true;
+            object_iter_data.idx_p                        = &idx_p;
+
+            if (!object_iter_data.iter_function.object_iter_op)
+                FUNC_GOTO_ERROR(H5E_OBJECT, H5E_BADVALUE, FAIL, "no object iteration function specified");
+
+            switch (loc_params->type) {
+                case (H5VL_OBJECT_BY_SELF): {
+
+                    /* Since the VOL doesn't directly pass down the group's hid_t, explicitly open the group
+                     * here so that a valid hid_t can be passed to the user's link iteration callback.
+                     */
+
+                    switch (loc_obj->obj_type) {
+                        case H5I_FILE:
+                        case H5I_GROUP:
+                            if (NULL == (iter_object = RV_group_open(loc_obj, loc_params, ".", H5P_DEFAULT,
+                                                                     H5P_DEFAULT, NULL)))
+                                FUNC_GOTO_ERROR(H5E_SYM, H5E_CANTOPENOBJ, FAIL,
+                                                "can't open object iteration group");
+
+                            parent_object_header = "groups";
+                            iter_object_type     = H5I_GROUP;
+                            break;
+
+                        case H5I_DATASET:
+                            if (NULL == (iter_object = RV_dataset_open(loc_obj, loc_params, ".", H5P_DEFAULT,
+                                                                       H5P_DEFAULT, NULL)))
+                                FUNC_GOTO_ERROR(H5E_SYM, H5E_CANTOPENOBJ, FAIL,
+                                                "can't open object iteration dataset");
+
+                            parent_object_header = "datasets";
+                            iter_object_type     = H5I_DATASET;
+                            break;
+
+                        case H5I_DATATYPE:
+                            if (NULL == (iter_object = RV_datatype_open(loc_obj, loc_params, ".", H5P_DEFAULT,
+                                                                        H5P_DEFAULT, NULL)))
+                                FUNC_GOTO_ERROR(H5E_SYM, H5E_CANTOPENOBJ, FAIL,
+                                                "can't open object iteration dataset");
+
+                            parent_object_header = "datatypes";
+                            iter_object_type     = H5I_DATATYPE;
+                            break;
+
+                        case H5I_ATTR: {
+                            /* The object where the attribute is attached will be iterated */
+                            switch (loc_obj->u.attribute.parent_obj_type) {
+                                case H5I_FILE:
+                                case H5I_GROUP:
+                                    if (NULL == (iter_object = RV_group_open(loc_obj, loc_params, ".",
```

Comment:
Have you tested H5Ovisit on an attribute? It doesn't seem like these open calls would be opening the correct object and would instead be trying to open the attribute object as a group, dataset, datatype, etc.

Reply:
Good catch - the API tests don't try H5Ovisit on an attribute. I'll implement a test for that case, and have this make a request to server to get the actual object the attribute is attached to.

Reply:
It would probably be a good idea to keep track of the object that an attribute is attached to at create or open time if we don't already rather than adding another call to the server here